### PR TITLE
ARROW-17455: [Go] Function and Kernel execution architecture

### PR DESCRIPTION
**Pull request by zeroshade**
In addition to implementing the function execution architecture, this also shifts some files around for better package naming and scoping. A new package `exec` is created inside of `compute/internal/` to make it easier to scope other internal-only methods / functionality such as the kernel implementations themselves.

This is a fairly large change because so much is interconnected, but it also includes extensive tests covering ~85% of the `internal/exec` package and nearly 80% of the entire `compute` package. More tests will be added as functionality is added.

### CODE REVIEW

File: go/arrow/compute/exec.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+				return nil, fmt.Errorf("%w: passed batch length for execution did not match actual length for scalar fn execution",
+					arrow.ErrInvalid)
+			}
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ch := make(chan Datum, 10)
```
Comment (cyb70289) [inline go/arrow/compute/exec.go:130]: `10` looks without strong support, but I don't have better suggestion.

Is channel strictly necessary here? I think channel make the code easier to read than C++ which collects all results with a datum accumulator.

**Review by cyb70289 — APPROVED** [review]
LGTM.

Will be great if there's a simple kernel to illustrate all pieces working together.

**Comment by zeroshade** [comment]
@cyb70289 you can see some simple kernels in the tests which are used to test the infrastructure :smile:

File: go/arrow/compute/exec.go (hunk 1 above)
Comment (zeroshade) [reply go/arrow/compute/exec.go:130]: You're correct that a channel is not strictly necessary here, I did it for the reason you suggested: it makes it more readable than the C++ using a datum accumulator.

I'll make the `10` a configurable setting via the `ExecCtx` and just leave 10 as the default for now. Ultimately i just wanted to make sure that neither the `Exec` nor the `WrapResults` got blocked in most cases on how quickly the execution works.

**Comment by zeroshade** [comment]
@cyb70289 also following this I'm already starting to implement kernels for various functions, so aside from the tests there'll also be lots of implemented kernels to use for examples.

**Pull request merged**